src/ir/access_path_root.cc

```diff
@@ -0,0 +1,34 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#include "src/ir/access_path_root.h"
+
+#include "src/ir/datalog_print_context.h"
+
+// The classes in this file describe the root of an AccessPath. At the moment,
+// we have only two types of roots: a HandleConnectionSpecAccessPathRoot and
+// a HandleConnectionAccessPathRoot, describing a HandleConnectionSpec in a
+// ParticleSpec and a Handle in a Particle, respectively.
+namespace raksha::ir {
+
+std::string AccessPathRoot::ToDatalog(const DatalogPrintContext &ctxt) const {
+  const auto *instantiation_map = ctxt.instantiation_map();
+  if (instantiation_map == nullptr) return this->ToString();
```

Comment:
Isn't this a case where we would like to assert via a `CHECK`? It seems most likely to be the case that if the `instantiation_map` is `nullptr` while `ToDatalog` is called that we intended to set it but forgot.

Reply:
Ideally, yes. However, there are tests where we cannot create a meaningful instantiation map as they already create the instantiated paths. Shoehorning a map in those tests will make them harder to read.